sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Listeners/ServiceBusListener.cs

```diff
@@ -656,55 +668,67 @@ private void CancelExistingMonitoringTasks()
                             ActivityKind.Consumer,
                             MessagingDiagnosticOperation.Process);
 
-            scope.SetMessageData(messagesArray);
-
-            scope.Start();
-            FunctionResult result = await _triggerExecutor.TryExecuteAsync(input.GetTriggerFunctionData(), _functionExecutionCancellationTokenSource.Token).ConfigureAwait(false);
-            if (result.Exception != null)
+            try
             {
-                scope.Failed(result.Exception);
-            }
-            receiveActions.EndExecutionScope();
+                scope.SetMessageData(messagesArray);
 
-            var processedMessages = messagesArray.Concat(receiveActions.Messages.Keys);
-            // Complete batch of messages only if the execution was successful
-            if (_autoCompleteMessages && result.Succeeded)
-            {
-                List<Task> completeTasks = new List<Task>();
-                foreach (ServiceBusReceivedMessage message in processedMessages)
+                scope.Start();
+                FunctionResult result = await _triggerExecutor
+                    .TryExecuteAsync(input.GetTriggerFunctionData(), _functionExecutionCancellationTokenSource.Token).ConfigureAwait(false);
+                if (result.Exception != null)
                 {
-                    // Skip messages that were settled in the user's function
-                    if (input.MessageActions.SettledMessages.ContainsKey(message))
-                    {
-                        continue;
-                    }
-
-                    // Pass CancellationToken.None to allow autocompletion to finish even when shutting down
-                    completeTasks.Add(receiver.CompleteMessageAsync(message, CancellationToken.None));
+                    scope.Failed(result.Exception);
                 }
 
-                await Task.WhenAll(completeTasks).ConfigureAwait(false);
-            }
-            else if (!result.Succeeded)
-            {
-                // For failed executions, we abandon the messages regardless of the autoCompleteMessages configuration.
-                // This matches the behavior that happens for single dispatch functions as the processor does the same thing
-                // in the Service Bus SDK.
+                receiveActions.EndExecutionScope();
 
-                List<Task> abandonTasks = new();
-                foreach (ServiceBusReceivedMessage message in processedMessages)
+                var processedMessages = messagesArray.Concat(receiveActions.Messages.Keys);
+                // Complete batch of messages only if the execution was successful
+                if (_autoCompleteMessages && result.Succeeded)
                 {
-                    // skip messages that were settled in the user's function
-                    if (input.MessageActions.SettledMessages.ContainsKey(message))
+                    List<Task> completeTasks = new List<Task>();
```

Comment:
```suggestion
                    List<Task> completeTasks = new List<Task>(messagesArray.Length + receiveActions.Messages.Keys.Count);
```

I doubt there's a measurable impact, but it seems we have the information readily available to right-size this.